_posts/2016-03-14-Containers-all-the-way-through.md

```diff
@@ -0,0 +1,95 @@
+canonical:
+    name: my personal blog
+    href: https://neomatrix369.wordpress.com/2016/03/13/containers-all-the-way-through/
+tags:
+- bare metal system
+- cgroups
+- container
+- container system
+- executable
+- hypervisor 
+- kernel
+- namespaces 
+- operating system
+- process 
+- shared operating system
+- shared resources
+- virtual hardware
+- virtual machine 
+- virtual system
+- vm
+--- 
+
+### Introduction
+
+In this post I will attempt to cover fundamentals of **Bare Metal Systems,** **Virtual Systems** and **Container Systems**. And the purpose for doing so is to learn about these systems as they stand and also the differences between them, focussing how they execute programs in their respective environments.
+
+### Bare metal systems
+
+Lets think of our bare metal systems as desktops and laptops we use on a daily basis, and we have the following components:
+
+*   The hardware (outer layer)
+*   The OS platform (running inside the hardware)
+*   programs running as processes on the OS
+
+Programs are stored on the hard drive in the form of executable files (a format understandable by the OS) and loaded into memory via one or more processes*. The programs interact with the kernel, which forms a core part of the OS architecture and the hardware. The OS coordinate communication between hardware i.e. CPU, I/O devices, etc… and programs.
+
+<img src="/assets/img/custom/blog/2016-03-14-containers-all-the-way-through/bare-metal-systems.png" alt="Bare Metal Systems" title="Bare Metal Systems" class="img img-center img-responsive style-screengrab">
+
+
+A more detailed explanation what programs or executables are, how programs execute and where does an Operating System come into play, can be found [here [2]](http://stackoverflow.com/questions/1599434/how-does-program-execute-where-does-the-operating-systems-come-into-play).
+
+### Virtual systems
+
+On the other hand virtual systems, with the help of virtual system controllers like, _Virtual Box_ or _VMWare_ or [_a_ _hypervisor [1]_](https://en.wikipedia.org/wiki/Hypervisor) to run an operating system running on the bare metal. These systems emulate bare-metal hardware as software abstraction(s) inside which we run the real OS platform. Such systems can be made up of such layers, and also referred to as Virtual Machines (VM):
+
+*   a software abstraction of the hardware (virtual machine)
+*   the OS platform running inside the software abstraction (guest OS)
+*   one or more programs running as processes in the guest OS
+
+Its like running a computer (abstracted as software) inside another computer. And the rest of the fundamentals from the bare metal system applies to this abstraction layer as well. When a process is created inside the virtual system, then the host OS which runs the virtual system might also be spawning one or more processes on the host OS in relation to this process.
```

Comment:
I have changed the sentence a bit, so it stays `host`, not talking about `guest` OS yet.